akka-distributed-data/src/main/scala/akka/cluster/ddata/ReplicatedData.scala

```diff
@@ -0,0 +1,82 @@
+/**
+ * Copyright (C) 2009-2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.cluster.ddata
+
+import akka.cluster.UniqueAddress
+
+/**
+ * Interface for implementing a state based convergent
+ * replicated data type (CvRDT).
+ *
+ * ReplicatedData types must be serializable with an Akka
+ * Serializer. It is highly recommended to implement a serializer with
+ * Protobuf or similar. The built in data types are marked with
+ * [[ReplicatedDataSerialization]] and serialized with
+ * [[akka.cluster.ddata.protobuf.ReplicatedDataSerializer]].
+ *
+ * Serialization of the data types are used in remote messages and also
+ * for creating message digests (SHA-1) to detect changes. Therefore it is
+ * important that the serialization produce the same bytes for the same content.
+ * For example sets and maps should be sorted deterministically in the serialization.
+ *
+ * ReplicatedData types should be immutable, i.e. "modifying" methods should return
+ * a new instance.
+ */
+trait ReplicatedData {
+  type T <: ReplicatedData
+
+  /**
+   * Monotonic merge function.
+   */
+  def merge(that: T): T
+
+}
+
+/**
+ * Java API: Interface for implementing a [[ReplicatedData]] in
+ * Java.
+ */
+abstract class AbstractReplicatedData extends ReplicatedData {
+  // it is not possible to use a more strict type, because it is erased somehow, and
+  // the implementation is anyway required to implement
+  // merge(that: ReplicatedData): ReplicatedData
+  type T = AbstractReplicatedData
+
+}
```

Comment:
There is a way to make this typesafe in Java, via "self recursive type parameters" (sometimes called F-bounds):

Here's some docs I did on the pattern: http://ktoso.github.io/scala-types-of-types/#self-recursive-type

```scala
class Data[T] extends Data[Data[T]] {
  def merge(d: T): T = ???
}
```

usage in java:
```java
  static class X extends Data<X> { }
  static class Y extends Data<Y> { }
  {
    X x = new X();
    Y y = new Y();

    x.merge(x);
    y.merge(y);
    x.merge(y); // good, compilation fails: merge(X) cannot be applied to Y
    y.merge(x); // good, compilation fails: merge(Y) cannot be applied to X
  }
```

So this could be improved in next versions for Java I think :-)